pkg/nuclide-ocaml/lib/OutlineProvider.js

```diff
@@ -0,0 +1,90 @@
+'use babel';
+/* @flow */
+
+/*
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the license found in the LICENSE file in
+ * the root directory of this source tree.
+ */
+
+import type {Outline} from '../../nuclide-outline-view';
+import type {MerlinProcess} from '../../nuclide-ocaml-base/lib/MerlinProcess';
+
+import {Point} from 'atom';
+
+import {
+  className,
+  method,
+  keyword,
+  plain,
+  type,
+  whitespace,
+} from '../../commons-node/tokenizedText';
+import {getServiceByNuclideUri} from '../../nuclide-remote-connection';
+
+function makeTokens(data) {
+  let kind;
+  let nameToken;
+
+  switch (data.kind) {
+    case 'Value':
+      kind = 'val';
+      nameToken = method(data.name);
+      break;
+    case 'Class':
+    case 'Exn':
+    case 'Module':
+      nameToken = className(data.name);
+      break;
+    case 'Constructor':
+      kind = 'ctor';
+      nameToken = className(data.name);
+      break;
+    case 'Signature':
+      kind = 'sig';
+      nameToken = className(data.name);
+      break;
+    case 'Type':
+      nameToken = type(data.name);
+      break;
+  }
+  kind = kind || data.kind.toLowerCase();
+  nameToken = nameToken || plain(data.name);
+
+  return [keyword(kind), whitespace(' '), nameToken];
+}
+
+function convertMerlinOutline(data) {
+  const tokenizedText = makeTokens(data);
+  const children = data.children.reverse().map(convertMerlinOutline);
+  const startPosition = new Point(data.start.line - 1, data.start.col);
+  const endPosition = new Point(data.end.line - 1, data.end.col);
+
+  return {
+    tokenizedText,
+    children,
+    startPosition,
+    endPosition,
+  };
+}
+
+export async function getOutline(editor: atom$TextEditor): Promise<?Outline> {
+  const path = editor.getPath();
+  if (path == null) {
+    return null;
+  }
+  const instance: ?MerlinProcess = getServiceByNuclideUri('MerlinService', path);
+  if (!instance) {
+    return null;
+  }
+  await instance.pushNewBuffer(path, editor.getText());
+  const result = await instance.outline(path);
+  if (!Array.isArray(result)) {
+    return null;
+  }
+  return {
+    outlineTrees: result.reverse().map(convertMerlinOutline),
```

Comment:
It's surprising that this needs to be reversed. I'm assuming that's the way merlin reports it? Could the `reverse` be pushed further down the stack to keep the higher-level APIs cleaner?